opentelemetry-api/src/opentelemetry/trace/span.py

```diff
@@ -75,7 +75,9 @@ def is_recording(self) -> bool:
         """
 
     @abc.abstractmethod
-    def set_status(self, status: Status) -> None:
+    def set_status(
+        self, status: typing.Union[str, StatusCode, Status]
```

Comment:
Accepting `str` would mean user could pass arbitrary string or different variations of enum value ex: `Error`, `error` or maybe even `err`. It raises `KeyError` in such cases with below SDK implementation and you also added test for that. [Error handling](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/error-handling.md#basic-error-handling-principles) insists _MUST NOT throw unhandled exceptions when used incorrectly by end users. The API and SDK SHOULD provide safe defaults for missing or invalid arguments._

As an end user If API is accepting `StatusCode` I assume it also accepts optional `description` and even [specification](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#set-status) says that. I think this change violates spec and doesn't add significant value.

Reply:
Didn't know error handling rule here, that was rather unexpected. I've changed code accordingly.

> doesn't add significant value.

I just brought pain points I experienced while I was using this lib.

Use of bare lib bloats application code a lot, it requires to add this boilerplate many times to many methods, sometimes multiple times per method. I believe that as instrumentation code is not a business logic code, it should be as thin as possible, ideally no code at all. Having that copypaste block definitely decreases readability of a main logic.

Reply:
I don't feel good about accepting arbitrary strings either. It sort of defeats the purpose of having enums in the first place. I think accepting an enum value or a status would probably be OK but then does it make enough of a difference? Probably not. 

Users can also define their own utility functions to reduce the boilerplate if it feels too verbose. Personally, I don't think it's a huge deal. The API as it stands today is actually very readable and obvious. I'd prioritize that over making it faster to type. People reading the code seeing three different styles of setting span status would just be confused.

Reply:
> Users can also define their own utility functions to reduce the boilerplate if it feels too verbose.

Doesn't that mean that lib is not really user-friendly, if people need to create helpers on top to make it more usable?

Of course I don't know how many people do that.

Reply:
When I am instrumenting my code I do this `span.set_status(Status(StausCode.ERROR))`. I agree with @owais on prioritising API over less boilerplate but confusing one. I think not allowing some arbitrary string to be passed is better than accepting it and surprising end users without doing anything. I feel the proposed change is less user-friendly than it is now. If I am coming to this library as a first time user I would have these question in mind with this change.
1. What does it mean to have both `Status` object and `description` parameters? Does it prefer the description from the object or description param?
2. Why does it have three different ways of setting status and how are they different from each other?
3. What is the ideal and recommended way out of the three alternative ways?
4. When `Error` or `error` string is given as status why doesn't it set `StatusCode.ERROR` status?

Reply:
> When I am instrumenting my code I do this `span.set_status(Status(StausCode.ERROR))`. 

Aren't there too many *status*es?


> I think not allowing some arbitrary string to be passed is better than accepting it and surprising end users without doing anything.

Didn't you just suggested [the opposite](https://github.com/open-telemetry/opentelemetry-python/pull/1518#discussion_r555060470)?


> If I am coming to this library as a first time user I would have these question in mind with this change.

Proposed solution isn't ideal of course but backward compatible. Another idea is to have methods:

```
def set_status_ok(self, description=None)
def set_status_error(self, description=None)
```
and not expose internal `Status` object at all.


Another note — you're talking about accepted types, string values and validation as it's somehow enforced now. In fact there is [no validation at all](https://github.com/open-telemetry/opentelemetry-python/blob/336af87a095deb46a0de8a9ef39b7f60c87e422c/opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py#L659). Mypy is not a validation.

Reply:
My initial comment was just to highlight this is spec incompatible change.

Reply:
I see

Reply:
Another approver here, and I agree with @lonewolf3739 and @owais that although this change is certainly well-intentioned and an improvement in terms of the number of characters one has to type, It doesn't improve the type of usability we want to promote for this library.

The opentelemetry specification focuses on uniformity for the purposes of improving the transferable knowledge of the API/SDKs, and the safety to reduce common mistakes. 

Accepting a string as a short hand for a type, although somewhat common in Python, tends to result in the run-time risks that @lonewolf3739 has raised. Eventually, large code bases tend to [err toward more type safety](https://dropbox.tech/application/our-journey-to-type-checking-4-million-lines-of-python) out of necessity to improve the ability for developers to ramp up.

I could be convinced that allowing a StatusCode in a method is fine, as that still can be caught by a type checker rather easily. But that also raises a poor practice of not clarifying the reason for the status with a string. 

For the specific example you raised in the PR description, you could make that a bit cleaner by enabling the import of both StatusCode and Status from trace:

```
from opentelemetry.trace import StatusCode, Status

span.set_status(Status(StatusCode.OK))
```

Taking a quick look, we can't import StatusCode from opentelemetry.trace yet. That would be a good addition that doesn't break the spec and reduces the modules one has to import from.

Reply:
Typechecking is not the thing you can fully rely on, n many cases it's a self-deception.

One argument:
```bash
opentelemetry-python$ grep -lr 'type: ignore'
opentelemetry-sdk/src/opentelemetry/sdk/trace/export/__init__.py
opentelemetry-sdk/tests/configuration/test_configurator.py
opentelemetry-instrumentation/src/opentelemetry/instrumentation/instrumentor.py
opentelemetry-instrumentation/src/opentelemetry/instrumentation/configurator.py
opentelemetry-instrumentation/src/opentelemetry/instrumentation/metric.py
opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/sitecustomize.py
opentelemetry-instrumentation/tests/test_bootstrap.py
opentelemetry-instrumentation/tests/test_instrumentor.py
opentelemetry-instrumentation/tests/test_metric.py
opentelemetry-instrumentation/tests/test_run.py
opentelemetry-api/src/opentelemetry/configuration/__init__.py
opentelemetry-api/src/opentelemetry/trace/__init__.py
opentelemetry-api/src/opentelemetry/context/contextvars_context.py
opentelemetry-api/src/opentelemetry/context/__init__.py
opentelemetry-api/src/opentelemetry/context/aiocontextvarsfix.py
opentelemetry-api/src/opentelemetry/propagators/__init__.py
opentelemetry-api/src/opentelemetry/propagators/composite.py
opentelemetry-api/tests/configuration/test_configuration.py
opentelemetry-api/tests/trace/test_status.py
opentelemetry-api/tests/trace/test_immutablespancontext.py
opentelemetry-proto/tests/test_proto.py
```

I wouldn't rely on type checking only.

On this basis I don't really think that `span.set_status('EGGOG')` error is more likely than one of:
```py
span.set_status(Status(StatusCode.Okayish))
span.set_status(StatusCode(Status.OK))
span.set_status(Status(StatusCode))
span.set_status(StatusCode.OK)
span.set_status("Any arbitrary object")
```
Especially when incorrect usage is [silently ignored](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/error-handling.md#basic-error-handling-principles).

Ok, I got that you don't like that approach — it's your project and you know best.

But what do you think of another way:
```py
def set_status_ok(self, description=None)
def set_status_error(self, description=None)
```
?

Reply:
Just in case if you are not already aware, The language implementation of API/SDK must follow the [Opentelemetry Specification](https://github.com/open-telemetry/opentelemetry-specification). The alternative way doesn't adhere to specs. 

> Especially when incorrect usage is silently ignored.

Opentelemetry believes users would rather lose the telemetry than instrumentation introducing errors; this is the reason API/SDK MUST NOT throw unhandled exceptions, invalid inputs are dropped and default values are set wherever possible etc... and are always _logged_ to make sure they are not silently ignored.

Reply:
```
Opentelemetry believes users would rather lose the telemetry than instrumentation introducing errors; this is the reason API/SDK MUST NOT throw unhandled exceptions, invalid inputs are dropped and default values are set wherever possible etc... and are always logged to make sure they are not silently ignored.
```

One other addition to this point: although we try our best to eliminate runtime errors, we generally have not gone out of the way to try to eat core syntactical errors.

In your examples above,  we don't do any special runtime handling to catch those. Many would manifest an a logged exception as @lonewolf3739 pointed out. some would result in an attribute error, like your first example. 

It might be good for opentelemetry-python to clarify it's stance here, but we don't catch an attributeerror from your first example.

@anton-ryzhov to the rest of your comment:

Yes, it's true that type checking nor this libraries runtimes are foolproof. But I don't see why an imperfect approach means we should enable an API that is even more imperfect.

At least type checking is a common tool, and anyone using it will get feedback on all 5 of the error cases you raised in their dev environment. I don't know of any tool that would catch an erroneous string that is intended to be interpreted as an enum. 

One other follow up question: there are clear negatives to exposing such an API. 

Even if there wasn't a negative to this particular function, there's always an inherent negative in introducing another way to do the same thing: it bifurcates examples and requires readers to know about the multiple ways of doing it.

Given the negatives, is the value of reducing two imports worth it?

Reply:
@lonewolf3739,

> Just in case if you are not already aware, The language implementation of API/SDK must follow the Opentelemetry Specification. The alternative way doesn't adhere to specs.

Thank you. Yes, I don't fully know your dev process. I was expecting someone to walk me through the extra steps if we agreed on something.


According to the [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#set-status):

> An API to set the Status. This SHOULD be called SetStatus. This API takes the **StatusCode**, and an optional Description, either **as individual parameters** or as an immutable object encapsulating them, **whichever is most appropriate for the language**.

In my POV `span.set_status(StatusCode.OK))` fits the spec even better than current `span.set_status(Status(StatusCode.OK))`. Since that wrapper is not required by the spec nor by the language. And even goes against "Simple is better than complex / Flat is better than nested / Readability counts" Zen.

Reply:
@toumorokoshi,

> One other addition to this point: although we try our best to eliminate runtime errors, we generally have not gone out of the way to try to eat core syntactical errors.

That is a reasonable approach, I totally get it.

But I don't see how this PR (original version) changes that anyhow.

Let's compare original `span.set_status(Status(StatusCode.Oak))` and suggested `span.set_status('Ouch')` in terms of validation ability and potential runtime errors.

 1. If we rely on typing. `mypy` will catch both cases if argument's type is defined as `typing.Literal['OK', 'ERROR', 'UNSET']` instead of just `str`. It even handles more complex cases like `span.set_status('OK' if success else 'ErroR')` by checking both options.

 2. Without type checks we have one "happy" (the best that python can offer) case — if that exact **line** will be executed on dev/test/stage environment. Both will crash with `AttributeError`/[`KeyError`](https://github.com/open-telemetry/opentelemetry-python/pull/1518/commits/9fcb80ab6f222fcc26e2f14a6a5a9a2824a7ab35#diff-5b32f7fdcee84637d5db9d21bbbdd33a4666ff56a683fd1ebd3651c5aa4596b2R780-R782).

 3. Or unhappy case — when it's only executed on prod. Both will crash again. "Sorry for that, it's python, increase test coverage".

I don't think passing a string anyhow worse/error-prone than using an Enum.



According to the [Spec](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#set-status) again:

> StatusCode is one of the following values

It's not said that it must be an Enum or some other object, is it? At least I read it that `StatusCode` could be a string from given list.


> There's always an inherent negative in introducing another way to do the same thing: it bifurcates examples and requires readers to know about the multiple ways of doing it.

Python's Zen agrees with you:

> There should be one-- and preferably only one --obvious way to do it.

But I want to emphasise **obvious way** there. For me `span.set_status('ERROR')` could be the only obvious way to do it. Of course keep the old one for backward compatibility, but don't advertise it anyhow.

Reply:
> But I want to emphasise obvious way there. For me span.set_status('ERROR') could be the only obvious way to do it. Of course keep the old one for backward compatibility, but don't advertise it anyhow.

This emphasizes another concern for me: we have 4 people who have spoken up on this review. three have said they do not like the design change. If a subjective meaning of 'obvious' is the primary argument, I'd err on the side of the majority (two of which are approvers). So I think there has to be a stronger argument than that to make the case here.

Despite our disagreement here, you are certainly entitled to your own opinion, and are welcome to continue advocating for your perspective. And I see you have other PRs too, and any contribution is always greatly appreciated! Thank you.

But I do want to drive toward a conclusion.

I won't approve this PR, unless there's some alternate argument. I won't mark as "changes requested" in case others change their minds on this PR. 

Also, this may be time for a maintainer to weigh in. pinging @codeboten and @lzchen.

Reply:
I know the rules. I gave my arguments, and it's up to the maintainers to decide.
So feel free to reject & close

Reply:
Given the number of votes against the change, I believe we should not move forward with this. Perhaps in the future if there are valid customer concerns about the usability of our API, we can ADD convenience functions but as of now I believe it is best to leave the api as it is.